prow/plugins/require-matching-label/require-matching-label.go

```diff
@@ -0,0 +1,213 @@
+			},
+		},
+		nil
+}
+
+type event struct {
+	org    string
+	repo   string
+	number int
+	author string
+	// The PR's base branch. If empty this is an Issue, not a PR.
+	branch string
+	// The label that was added or removed. If empty this is an open or reopen event.
+	label string
+	// The labels currently on the issue. For PRs this is not contained in the webhook payload and may be omitted.
+	currentLabels []github.Label
+}
+
+func handleIssue(pc plugins.PluginClient, ie github.IssueEvent) error {
+	if !handleIssueActions[ie.Action] {
+		return nil
+	}
+	e := &event{
+		org:           ie.Repo.Owner.Login,
+		repo:          ie.Repo.Name,
+		number:        ie.Issue.Number,
+		author:        ie.Issue.User.Login,
+		label:         ie.Label.Name, // This will be empty for non-label events.
+		currentLabels: ie.Issue.Labels,
+	}
+	return handle(pc.Logger, pc.GitHubClient, pc.CommentPruner, pc.PluginConfig.RequireMatchingLabel, e)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if !handlePRActions[pre.Action] {
+		return nil
+	}
+	e := &event{
+		org:    pre.Repo.Owner.Login,
+		repo:   pre.Repo.Name,
+		number: pre.PullRequest.Number,
+		author: pre.PullRequest.User.Login,
+		label:  pre.Label.Name, // This will be empty for non-label events.
+	}
+	return handle(pc.Logger, pc.GitHubClient, pc.CommentPruner, pc.PluginConfig.RequireMatchingLabel, e)
+}
+
+// matchingConfigs filters irrelevant RequireMtchingLabel configs from
+// the list of all configs.
+// `branch` should be empty for Issues and non-empty for PRs.
+// `label` should be omitted in the case of 'open' and 'reopen' actions.
+func matchingConfigs(org, repo, branch, label string, allConfigs []plugins.RequireMatchingLabel) []plugins.RequireMatchingLabel {
+	var filtered []plugins.RequireMatchingLabel
+	for _, cfg := range allConfigs {
+		// Check if the config applies to this issue type.
+		if (branch == "" && !cfg.Issues) || (branch != "" && !cfg.PRs) {
+			continue
+		}
+		// Check if the config applies to this 'org[/repo][/branch]'.
+		if org != cfg.Org ||
+			(cfg.Repo != "" && cfg.Repo != repo) ||
+			(cfg.Branch != "" && branch != "" && cfg.Branch != branch) {
+			continue
+		}
+		// If we are reacting to a label event, see if it is relevant.
+		if label != "" && !cfg.Re.MatchString(label) {
+			continue
+		}
+		filtered = append(filtered, cfg)
+	}
+	return filtered
+}
+
+func handle(log *logrus.Entry, ghc githubClient, cp commentPruner, configs []plugins.RequireMatchingLabel, e *event) error {
+	// Find any configs that may be relevant to this event.
+	matchConfigs := matchingConfigs(e.org, e.repo, e.branch, e.label, configs)
+	if len(matchConfigs) == 0 {
+		return nil
+	}
+
+	if e.label == "" /* not a label event */ {
+		// If we are reacting to a PR or Issue being created or reopened, we should wait a
+		// few seconds to allow other automation to apply labels in order to minimize thrashing.
+		sleep(time.Second * 5)
```

Comment:
oh boy arbitrary hard-coded sleep :grimacing:

Reply:
What do you suggest? We don't have a mechanism to detect when automation has finished handling an event.
We could get rid of this, but it will cause the plugin to add and remove the missing label every time an issue or PR is opened with a matching label.

Reply:
do we have an estimate of how long it *needs* to be?
if this were another tool, I'd probably make it possible to override, because different deployments have different latency :grimacing:

Reply:
> do we have an estimate of how long it needs to be?

Nothing more specific than 'a few seconds'. We need to wait for any other automation that may be related to have time to react to the event, but we don't have an SLA on time-to-respond to GitHub events.

I'll add a configurable override for this duration, but I doubt it will ever be used, mainly because the failure case isn't actually problematic (it just looks strange and wastes a couple API tokens).

Reply:
Done.